DCOS-56537: add data attributes to simplify selectors for Cypress testing

mperrotti opened the pull request.

mperrotti commented:
Reviewers: This PR isn't as big and scary as it looks - it's all simple changes :)

GeorgiSTodorov commented on packages/appChrome/components/SidebarItem.tsx:
```diff
@@ -14,6 +14,11 @@ export interface SidebarItemProps {
 class SidebarItemComponent extends React.PureComponent<SidebarItemProps, {}> {
   public render() {
     const { children, isActive, onClick } = this.props;
+    const dataCy = [
```
I can't find the data-cy for `SidebarItem`s.

pierrebeitz replied:
@GeorgiSTodorov 	is right. we need to search for things that are unwrapped. e.g. `Clickable` will be unwrapped, so only it's children are actually rendered. thus we lose our `data-cy`-prop.

GeorgiSTodorov commented on packages/appChrome/components/SidebarSubMenuItem.tsx:
```diff
@@ -64,9 +64,13 @@ class SidebarSubMenuItem extends React.PureComponent<
         )]: isActive
       }
     );
+    const dataCy = [
```
I can't see the data-cy for this element.

pierrebeitz replied:
works for me!

![Storybook 2019-07-22 10-49-05](https://user-images.githubusercontent.com/300861/61619035-5beab800-ac6e-11e9-8a6d-273f79a585c8.png)

GeorgiSTodorov commented on packages/badge/components/badgeButton.tsx:
```diff
@@ -40,7 +40,7 @@ export class BadgeButton extends React.PureComponent<BadgeButtonProps, {}> {
     `;
     return (
-      <Clickable {...props}>
+      <Clickable {...props} data-cy="badgeButton">
```
I can't see the data-cy for this element.

GeorgiSTodorov commented on packages/breadcrumb/components/Breadcrumb.tsx:
```diff
@@ -37,6 +37,7 @@ class Breadcrumb extends React.PureComponent<BreadcrumbProps, {}> {
             className={cx(textWeight("medium"), textSize("l"), {
               [padding("left", "xs")]: i !== 0
             })}
+            data-cy="breadcrumb"
```
If we have One > Two, then "One", ">" and "Two" all have this data-cy. Is this expected?

mperrotti replied:
That is expected since each of those is a breadcrumb. Do you (or any other reviewers) think we should exclude `data-cy` on the ">" breadcrumbs?

GeorgiSTodorov replied:
I don't have a strong opinion, I just though it was a little strange. If I don't know this, I would think that the second breadcrumb would be "Two", not ">".

GeorgiSTodorov commented on packages/clickable/components/clickable.tsx:
```diff
@@ -23,6 +23,10 @@ export interface ClickableProps {
+  /**
+   * human-readable selector used for writing tests
+   */
+  dataCy?: string;
```
This element doesn't have a data-cy.

mperrotti replied:
It's not supposed to. `Clickable` just renders it's children, so if you're using it on it's own, you just have to pass the `dataCy` prop

pierrebeitz commented:
i'm totally in favour of having this as it will make our tests way better!  

the only concern that came to my mind was the bundle size. but as that does not accumulate like it would in static HTML but only adds a couple bytes per component, we won't see a noticeable increase, right?

GeorgiSTodorov commented on packages/table/components/CheckboxTable.tsx:
```diff
@@ -154,7 +154,11 @@ class CheckboxTable extends React.PureComponent<
 
     return (
-      <div ref={this.checkboxRef} className={display("inline-block")}>
+      <div
```
I can't see the data-cy for this element.

mperrotti replied:
Really? I do
<img width="591" alt="Screen Shot 2019-07-19 at 10 44 40 AM" src="https://user-images.githubusercontent.com/2313998/61543856-4416f800-aa12-11e9-906d-f68a8d2654c3.png">

GeorgiSTodorov replied:
Which element did you select?
![Screenshot from 2019-07-19 17-51-59](https://user-images.githubusercontent.com/40791275/61544422-18b1fe80-aa4e-11e9-94e4-c22a413b6f5e.png)

mperrotti replied:
Sorry, my screenshots weren't very helpful :x

<img width="612" alt="Screen Shot 2019-07-19 at 10 58 12 AM" src="https://user-images.githubusercontent.com/2313998/61544817-2d71a080-aa14-11e9-9ebf-db0b7fa3ab14.png">

That `data-cy` is just for the checkbox in the table header

GeorgiSTodorov replied:
What if we want a table row?

mperrotti replied:
You could use the checkbox id or select using the `data-cy` attributes on the parent cell, but I've added a data-cy attribute specifically for checkbox cells for simplicity: `table-checkboxCell table-checkboxCell.row${rowIndex}`

GeorgiSTodorov commented on packages/table/components/Table.tsx:
```diff
@@ -421,6 +421,11 @@ export class Table<T> extends React.PureComponent<TableProps, TableState> {
     const onStartCallback = () => {
       this.setState({ resizeIndex: columnIndex });
     };
+    const dataCy = [
```
I can't see the data-cy for this element.

mperrotti replied:
I'm seeing it:
<img width="781" alt="Screen Shot 2019-07-19 at 10 45 48 AM" src="https://user-images.githubusercontent.com/2313998/61543962-6f014c00-aa12-11e9-9825-49e5fbb2c565.png">

GeorgiSTodorov replied:
This is just for the header?

mperrotti replied:
Yes

GeorgiSTodorov commented on packages/textInput/components/TextInput.tsx:
```diff
@@ -75,12 +75,20 @@ export class TextInput<
-        <div className={flex({ align: "center" })}>
+      <div {...containerProps} data-cy={dataCy}>
+        <div className={flex({ align: "center" })} data-cy="textInput-label">
           {labelContent}
```
Is there a reason why we don't have a data-cy for the actual input here? I think it's the most used part of this component.

mperrotti replied:
That's a great point. I'll add it

GeorgiSTodorov commented on packages/textarea/components/Textarea.tsx:
```diff
@@ -98,8 +105,14 @@ class Textarea extends React.PureComponent<TextareaProps, {}> {
+            {getHintContent ||
+              (hasError &&
+                getValidationErrors && (
+                  <div data-cy="textarea-hintContent">
```
I can't see the data-cy for hints.

GeorgiSTodorov commented on packages/tooltip/components/Tooltip.tsx:
```diff
@@ -58,6 +58,7 @@ class Tooltip extends React.PureComponent<TooltipProps, TooltipState> {
         onFocus={this.handleOpen}
         onBlur={this.handleClose}
         tabIndex={0}
+        data-cy="tooltip"
```
I can't see the data-cy for this element.

mperrotti replied:
I do
<img width="543" alt="Screen Shot 2019-07-19 at 10 56 07 AM" src="https://user-images.githubusercontent.com/2313998/61544632-dd92d980-aa13-11e9-9470-926a2b0bb234.png">

pierrebeitz commented on README.md:
```diff
@@ -93,6 +93,21 @@ describe('Badge', () => {
+
+**Child nodes:** If a child node has a `data-cy` added, there will be a dash between the parent node's name and a string to describe the child node. For example: The footer element of a `<DialogModal>` will have `data-cy="fullscreenModal-footer"`
+
+**States and variants:** If a node has a special "state", `data-cy` will prepend a string describing that state after a dot.
```
# 😻

mperrotti commented:
@pierrebeitz - I had a similar concern about bundle size, but it will be a very small increase

mperrotti commented:
I addressed @GeorgiSTodorov's feedback. In response to [this comment](https://github.com/dcos-labs/ui-kit/pull/379#discussion_r304819856), I also added a `data-cy` to each <input> (or <textarea> or <select>) element. This feels a little redundant to me, but might make writing selectors for form elements easier.

Reviewers: let me know what you think

pierrebeitz commented on packages/infobox/components/InfoBox.tsx:
```diff
@@ -67,10 +67,14 @@ export class InfoBox extends React.PureComponent<InfoBoxProps, {}> {
     const isUrgentMessage = appearance === "danger" || appearance === "warning";
+    const dataCy = [
+      "infoBox",
+      ...(appearance && appearance !== "default"
```
saw this pattern a couple of times now and realized that i would have lazily gone with having `appearance.default` and justifying it to myself with the slightly lighter footprint on implementation cost (yes, the conditional). 

no action item, just wanted to share
